src/index.ts

```diff
@@ -52,59 +54,66 @@ export function doubleCsrf({
     code: code,
   });
 
-  const generateTokenAndHash = (
+  const constructMessage = (req: Request, randomValue: string) => {
+    const uniqueIdentifier = getSessionIdentifier(req);
+    const messageValues = [
+      uniqueIdentifier.length,
+      uniqueIdentifier,
+      randomValue.length,
+      randomValue,
+    ];
+    return messageValues.join(messageDelimiter);
+  };
+
+  const getPossibleSecrets = (req: Request) => {
+    const getSecretResult = getSecret(req);
+    return Array.isArray(getSecretResult) ? getSecretResult : [getSecretResult];
+  };
+
+  const generateHmac = (secret: string, message: string) => {
+    return createHmac(hmacAlgorithm, secret).update(message).digest("hex");
+  };
+
+  const generateCsrfTokenInternal = (
     req: Request,
     {
       overwrite,
       validateOnReuse,
     }: Omit<GenerateCsrfTokenConfig, "cookieOptions">,
   ) => {
-    const getSecretResult = getSecret(req);
-    const possibleSecrets = Array.isArray(getSecretResult)
-      ? getSecretResult
-      : [getSecretResult];
-
-    const csrfCookie = getCsrfCookieFromRequest(req);
+    const possibleSecrets = getPossibleSecrets(req);
+    // If cookie is not present, this is a new user (no existing csrfToken)
     // If ovewrite is true, always generate a new token.
     // If overwrite is false and there is no existing token, generate a new token.
     // If overwrite is false and there is an existin token then validate the token and hash pair
     // the existing cookie and reuse it if it is valid. If it isn't valid, then either throw or
     // generate a new token based on validateOnReuse.
-    if (typeof csrfCookie === "string" && !overwrite) {
-      const [csrfToken, csrfTokenHash] = csrfCookie.split(delimiter);
-      if (
-        validateTokenAndHashPair({
-          csrfToken,
-          csrfTokenHash,
-          possibleSecrets,
-          sessionIdentifier: getSessionIdentifier(req),
-        })
-      ) {
-        // If the pair is valid, reuse it
-        return { csrfToken, csrfTokenHash };
+    if (cookieName in req.cookies && !overwrite) {
```

Comment:
Technically it should have already been checking if the cookie existed first, before getting the cookie out, this is just a more sane check. If the cookie isn't present, treat it as a new request.